### PR TITLE
Make all functionality invokable via the main Meson binary,

### PR DESCRIPTION
**Pull request by jpakkane**
which can be a Windows .exe file.

This is needed to make Meson installed as an MSI package work out of the box.

Probably will fail in the VS backend tests. If so, I'll fix those.

It would be cool to get this in 0.42.1, but it might be a bit too disruptive.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2190,9 +2191,8 @@ def func_vcs_tag(self, node, args, kwargs):
             else:
                 vcs_cmd = [' '] # executing this cmd will fail in vcstagger.py and force to use the fallback string
         # vcstagger.py parameters: infile, outfile, fallback, source_dir, replace_string, regex_selector, command...
-        kwargs['command'] = [sys.executable,
-                             self.environment.get_build_command(),
-                             '--internal',
+        kwargs['command'] = self.environment.get_build_command() + \
+                            ['--internal',
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/modules/gnome.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
 import sys
 import copy
 import subprocess
+import shlex
```
Comment (ignatenkobrain) [inline mesonbuild/modules/gnome.py]: __[flake8]__
* [F401] 'shlex' imported but unused

File: mesonbuild/backend/ninjabackend.py [hunk 3]
```diff
@@ -594,12 +592,11 @@ def generate_run_target(self, target, outfile):
 
     def generate_coverage_rules(self, outfile):
         e = NinjaBuildElement(self.all_outputs, 'meson-coverage', 'CUSTOM_COMMAND', 'PHONY')
-        e.add_item('COMMAND', [sys.executable,
-                               self.environment.get_build_command(),
-                               '--internal', 'coverage',
-                               self.environment.get_source_dir(),
-                               self.environment.get_build_dir(),
-                               self.environment.get_log_dir()])
+        e.add_item('COMMAND', self.environment.get_build_command() +
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [W291] trailing whitespace

File: mesonbuild/backend/ninjabackend.py [hunk 4]
```diff
@@ -896,13 +892,12 @@ def generate_rules(self, outfile):
...
         outfile.write('rule REGENERATE_BUILD\n')
-        c = (ninja_quote(quote_func(sys.executable)),
-             ninja_quote(quote_func(self.environment.get_build_command())),
-             '--internal',
-             'regenerate',
-             ninja_quote(quote_func(self.environment.get_source_dir())),
-             ninja_quote(quote_func(self.environment.get_build_dir())))
-        outfile.write(" command = %s %s %s %s %s %s --backend ninja\n" % c)
+        c = [ninja_quote(quote_func(x)) for x in self.environment.get_build_command()] + \
+             ['--internal',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/ninjabackend.py [hunk 5]
```diff
@@ -1500,13 +1495,13 @@ def generate_dynamic_link_rules(self, outfile):
                 outfile.write(description)
                 outfile.write('\n')
         outfile.write('\n')
+        args = [ninja_quote(quote_func(x)) for x in self.environment.get_build_command()] + [
+                '--internal',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E126] continuation line over-indented for hanging indent

File: mesonbuild/backend/ninjabackend.py [hunk 6]
```diff
@@ -2530,31 +2522,27 @@ def get_user_option_args(self):
...
         elem.add_item('DESC', 'Creating source packages')
-        elem.add_item('COMMAND', [sys.executable,
-                                  self.environment.get_build_command(),
-                                  '--internal', 'dist',
-                                  self.environment.source_dir,
-                                  self.environment.build_dir,
-                                  sys.executable,
-                                  self.environment.get_build_command()])
+        elem.add_item('COMMAND', self.environment.get_build_command() +
+                                  ['--internal', 'dist',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/ninjabackend.py [hunk 7]
```diff
@@ -2530,31 +2522,27 @@ def get_user_option_args(self):
...
         # Alias that runs the target defined above
         self.create_target_alias('meson-dist', outfile)
 
     # For things like scan-build and other helper tools we might have.
     def generate_utils(self, outfile):
-        cmd = [sys.executable, self.environment.get_build_command(),
-               '--internal', 'scanbuild', self.environment.source_dir, self.environment.build_dir,
-               sys.executable, self.environment.get_build_command()] + self.get_user_option_args()
+        cmd = self.environment.get_build_command() + \
+               ['--internal', 'scanbuild', self.environment.source_dir, self.environment.build_dir] + \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E126] continuation line over-indented for hanging indent

File: mesonbuild/interpreter.py [hunk 8]
```diff
@@ -2190,9 +2197,8 @@ def func_vcs_tag(self, node, args, kwargs):
...
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/vs2010backend.py [hunk 9]
```diff
@@ -413,8 +413,8 @@ def gen_run_target_vcxproj(self, target, ofname, guid):
         cmd = [sys.executable, os.path.join(self.environment.get_script_dir(), 'commandrunner.py'),
                self.environment.get_build_dir(),
                self.environment.get_source_dir(),
-               self.get_target_dir(target),
-               get_meson_script(self.environment, 'mesonintrospect')]
+               self.get_target_dir(target)] + \
+               self.environment.get_build_command() + ['introspect']
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/vs2010backend.py [hunk 10]
```diff
@@ -1093,9 +1092,8 @@ def gen_regenproj(self, project_name, ofname):
         ET.SubElement(midl, 'TypeLibraryName').text = '%(Filename).tlb'
         ET.SubElement(midl, 'InterfaceIdentifierFilename').text = '%(Filename)_i.c'
         ET.SubElement(midl, 'ProxyFileName').text = '%(Filename)_p.c'
-        regen_command = [sys.executable,
-                         self.environment.get_build_command(),
-                         '--internal',
+        regen_command = self.environment.get_build_command() + \
+                         ['--internal',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/vs2010backend.py [hunk 11]
```diff
@@ -1176,9 +1174,8 @@ def gen_testproj(self, target_name, ofname):
         postbuild = ET.SubElement(action, 'PostBuildEvent')
         ET.SubElement(postbuild, 'Message')
         # FIXME: No benchmarks?
-        test_command = [sys.executable,
-                        self.environment.get_build_command(),
-                        'test',
+        test_command = self.environment.get_build_command() + \
+                        ['test',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2010backend.py]: __[flake8]__
* [E127] continuation line over-indented for visual indent

File: mesonbuild/backend/ninjabackend.py [hunk 12]
```diff
@@ -2530,31 +2522,27 @@ def get_user_option_args(self):
...
         elem.add_item('DESC', 'Creating source packages')
-        elem.add_item('COMMAND', [sys.executable,
-                                  self.environment.get_build_command(),
-                                  '--internal', 'dist',
-                                  self.environment.source_dir,
-                                  self.environment.build_dir,
-                                  sys.executable,
-                                  self.environment.get_build_command()])
+        elem.add_item('COMMAND', self.environment.get_build_command() +
+            ['--internal', 'dist',
```
Comment (ignatenkobrain) [inline mesonbuild/backend/ninjabackend.py]: __[flake8]__
* [E128] continuation line under-indented for visual indent

**Comment by nirbheek** [comment]
It's a bit risky to merge this, but I think it's worth it overall. We can do another point release to fix regressions, if needed.

**Pull request merged**